### PR TITLE
Implement etag support for category files

### LINKED ISSUE #1653: Improve `hacsfiles` caching
One of my installs loads slow on mobile and I did some investigating and narrowed part of the issue  down to the frontend having to re-download 9 js files in `/hacsfiles/...`

I'm trying to come up with a way to solve caching issue with the frontend files being loaded every time because of the `custom_components.hacs.webresponses.category.async_serve_category_file` explicitly disabling caching which I assume is to ensure updates can happen without clearing browser cache.

```
                response.headers["Cache-Control"] = "no-store, max-age=0"
                response.headers["Pragma"] = "no-store"
```

In my install I have 9 included JS files like:

`/hacsfiles/lovelace-card-tools/card-tools.js`

I'm thinking one way to allow caching would be do so something like

`/hacsfiles/id-<HASH>/lovelace-card-tools/card-tools.js`

Then any url that matches `/hacsfiles/id-` could have `id-<HASH>` stripped.

When the `card-tools.js` file changes, recalculate the hash of the file contents and change the url in the frontend if needed.

### PR DESCRIPTION
**Pull request by bdraco**
Since I couldn't come with a solution that avoided checking the files, we can at least prevent re-downloading them if they haven't changed which means refreshing the mobile app has to download ~160 bytes per file instead of the whole file which still makes a big difference.

Retains the `max-age=0` and forces revalidation.

Fixes #1653

```
2020-11-16 05:12:31 DEBUG (MainThread) [custom_components.hacs] Serving simple-thermostat/simple-thermostat.js from /config/www/community/simple-thermostat/simple-thermostat.js with etag 0x6cda4805 (not-modified)
2020-11-16 05:12:31 DEBUG (MainThread) [custom_components.hacs] Serving lovelace-fold-entity-row/fold-entity-row.js from /config/www/community/lovelace-fold-entity-row/fold-entity-row.js with etag 0xd548b9d3 (not-modified)
2020-11-16 05:12:31 DEBUG (MainThread) [custom_components.hacs] Serving lovelace-card-tools/card-tools.js from /config/www/community/lovelace-card-tools/card-tools.js with etag 0x444c2ffc (not-modified)
2020-11-16 05:12:31 DEBUG (MainThread) [custom_components.hacs] Serving lovelace-slider-entity-row/slider-entity-row.js from /config/www/community/lovelace-slider-entity-row/slider-entity-row.js with etag 0x1c2fe06c (not-modified)
2020-11-16 05:12:31 DEBUG (MainThread) [custom_components.hacs] Serving garbage-collection-card/garbage-collection-card.js from /config/www/community/garbage-collection-card/garbage-collection-card.js with etag 0x3fd15b8d (not-modified)
2020-11-16 05:12:31 DEBUG (MainThread) [custom_components.hacs] Serving mini-graph-card/mini-graph-card-bundle.js from /config/www/community/mini-graph-card/mini-graph-card-bundle.js with etag 0x9d487a51 (not-modified)
2020-11-16 05:12:31 DEBUG (MainThread) [custom_components.hacs] Serving search-card/search-card.js from /config/www/community/search-card/search-card.js with etag 0xb87350ef (not-modified)
2020-11-16 05:12:31 DEBUG (MainThread) [custom_components.hacs] Serving battery-state-card/battery-state-card.js from /config/www/community/battery-state-card/battery-state-card.js with etag 0xd0b41886 (not-modified)
2020-11-16 05:12:31 DEBUG (MainThread) [custom_components.hacs] Serving lovelace-card-templater/lovelace-card-templater.js from /config/www/community/lovelace-card-templater/lovelace-card-templater.js with etag 0xf30f0e11 (not-modified)
```

### CODE REVIEW

File: custom_components/hacs/manifest.json [hunk 1]
```diff
@@ -20,6 +20,7 @@
     "backoff>=1.10.0",
     "hacs_frontend==202011151903",
     "semantic_version>=2.8.5",
-    "queueman==0.5"
+    "queueman==0.5",
+    "fnvhash==0.1.0"
```
Comment (bdraco) [inline custom_components/hacs/manifest.json:24]: Already used in `homekit`

**Comment by bdraco** [comment]
Had an opportunity to production test this since lovelace card tools updated and verified its working as expected

Screen Shot 2020-11-18 at 3 09 42 PM

**Review by ludeeus — APPROVED** [review]

File: custom_components/hacs/manifest.json [hunk 2]
```diff
@@ -18,8 +18,9 @@
     "aiofiles>=0.5.0",
     "aiogithubapi>=2.0.0<3.0.0",
     "backoff>=1.10.0",
+    "fnvhash==0.1.0",
```
Comment (ludeeus) [inline custom_components/hacs/manifest.json:21]:
```suggestion
    "fnvhash>=0.1.0<1.0.0",
```
@bdraco Should probably make this more forgiving incase it updates in core?

**Comment by quthla** [comment]
This massively slows down loading of the frontend by more than 5 secs when served by an Raspberry Pi 3.

Is the computation of the etag so heavy? Maybe just stick with sending 304 when received last modified equals to local last modified? Had to revert this for now and load time is again down to 1 sec.

**Comment by bdraco** [comment]
> This massively slows down loading of the frontend by more than 5 secs when served by an Raspberry Pi 3.
> 
> 
> 
> Is the computation of the etag so heavy? Maybe just stick with sending 304 when received last modified equals to local last modified? Had to revert this for now and load time is again down to 1 sec. 

Is your system cpu bound?  On my Rpi4, the time to hash the file is less than 1ms

**Comment by quthla** [comment]
Not at all. It is pretty idle.

load average: 0.40, 0.24, 0.27

**Comment by quthla** [comment]
I see you edited your message. Can you post a gist with the benchmarks you need so I can run it on my Pi?

**Comment by bdraco** [comment]
@quthla Its probably easier to try this PR: https://github.com/hacs/integration/pull/1744

**Comment by quthla** [comment]
Much better now but why are you enclosing the hash in quotes?

**Comment by bdraco** [comment]
> Can you post a gist with the benchmarks you need so I can run it on my Pi?
> 
> 
> 
> And you don't like the idea of just relying on last modified header? 

If-Modified-Since has trouble with going back in time when restoring from backups which makes hashing the file contents is generally preferred.

**Comment by bdraco** [comment]
> Much better now but why are you enclosing the hash in quotes?

Didn't work for me in Firefox until I did: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/ETag

**Comment by quthla** [comment]
It seems nginx won't cache the item if max age is set to 0.

**Comment by bdraco** [comment]
> It seems nginx won't cache the item if max age is set to 0.

That behavior is existing. Please consider sending a PR to address.

**Comment by quthla** [comment]
PR to where? Wouldn't it work if you set max age to something higher than 0?

**Comment by quthla** [comment]
https://trac.nginx.org/nginx/ticket/1182

There's a discussion about that behavior. Maybe worth setting the header mentioned?

**Pull request merged**